ctk/common/src/test/kotlin/org/codice/compilance/verification/core/CoreVerifierSpec.kt

```diff
@@ -0,0 +1,122 @@
+        every { specRefMsg.name } returns "TEST"
+
+        "single error status response has expected code" {
+            val statusCode = "urn:samlconf:fail"
+
+            buildDom(response(statusCode)).let {
+                CoreVerifier.verifyErrorStatusCode(it, specRefMsg, statusCode)
+            }
+        }
+
+        "mismatched status code fails and does not contain SAMLCore 3.2.1 response" {
+            buildDom(response("urn:samlconf:fail")).let {
+                shouldThrow<SAMLComplianceException> {
+                    CoreVerifier.verifyErrorStatusCode(it, specRefMsg, "urn:samlconf:expected")
+                }.message?.shouldNotContain(SAMLCore_3_2_1_d.message)
+            }
+        }
+
+        "failure with expected code of REQUESTER includes SAMLCore 3.2.1 response" {
+            buildDom(response("urn:samlconf:fail")).let {
+                shouldThrow<SAMLComplianceException> {
+                    CoreVerifier.verifyErrorStatusCode(it, specRefMsg, TestCommon.REQUESTER)
+                }.message?.shouldContain(SAMLCore_3_2_1_d.message)
+            }
+        }
+
+        "validateTimeWindow with null datetime attributes" {
+            buildDom("<fld>this is garbage</fld>").let {
+                CoreVerifier.validateTimeWindow(it, specRefMsg)
+            }
+
+            buildDom("""<fld NotBefore="$now"/>""").let {
+                CoreVerifier.validateTimeWindow(it, specRefMsg)
+            }
+
+            buildDom("""<fld NotOnOrAfter="$now"/>""").let {
+                CoreVerifier.validateTimeWindow(it, specRefMsg)
+            }
+
+            buildDom("""<fld NotBefore="$later"/>""").let {
+                CoreVerifier.validateTimeWindow(it, specRefMsg)
+            }
+
+            buildDom("""<fld NotOnOrAfter="$later"/>""").let {
+                CoreVerifier.validateTimeWindow(it, specRefMsg)
+            }
+        }
+
+        "validateTimeWindow with correct datetime attributes" {
+            buildDom("""<fld NotBefore="$now" NotOnOrAfter="$later"/>""").let {
+                CoreVerifier.validateTimeWindow(it, specRefMsg)
+            }
+        }
+
+        "validateTimeWindow with incorrect datetime attributes" {
+            buildDom("""<fld NotBefore="$later" NotOnOrAfter="$now"/>""").let {
+                shouldThrow<SAMLComplianceException> {
+                    CoreVerifier.validateTimeWindow(it, specRefMsg)
+                }
+            }
+
+            buildDom("""<fld NotBefore="$now" NotOnOrAfter="$now"/>""").let {
+                shouldThrow<SAMLComplianceException> {
+                    CoreVerifier.validateTimeWindow(it, specRefMsg)
+                }
+            }
+
+            buildDom("""<fld NotBefore="$later" NotOnOrAfter="$later"/>""").let {
+                shouldThrow<SAMLComplianceException> {
+                    CoreVerifier.validateTimeWindow(it, specRefMsg)
+                }
+            }
+        }
+    }
+}
```

Comment:
❗️ I get not verifying that the CoreVerifier delegates to the other Verifier objects. That sounds like a pain and would require significant changes. But shouldn't we be at least testing the other two pieces of functionality in this class? `preProcess` and `verifyEntityIdentifiers`? I'd be fine with making those functions internal to make that possible.

Reply:
`preProcess` basically tests:

1. Schema validation (outside the scope of this test class)
2. `verifyEncryptedElements`, which is implemented in the subclasses of `CoreVerifier` and should be tested against those subclasses
3. The functioning of the `EncryptionVerifier` (outside the scope of this test class)

Unless there's something else in there you think should be tested, I'm not sure how to dive on that.

`verifyEntityIdentifiers` is OBE, as it's been moved out of this class.

Reply:
My main concern is that we could have a bug in the flow of logic in the `preProcess` method. Or we could introduce one in the future. That being said, it would be a step above a purely "unit" test. Maybe the fact that we can't test this without calling out to a lot of external classes means the method should be moved or redesigned 🤔